### ISSUE TITLE
Core/Creature: Talk() in KilledUnit

### ISSUE THREAD

**Issue by offl** [TrinityCore/TrinityCore#26456]
We need a way to add cooldown for all those Talk(SAY_SLAY);
So after first Talk(SAY_SLAY); creature will be not able to say it again in next x seconds
For example verified CD is 10s and 20s

I think nearly all on kill texts has cooldown, so better to make it non-exclusive for BossAI only
Also I think it's good to allow to perform a check for player inside that function and not directly in KilledUnit, probably all creatures can say something only after killing player but who knows

```cpp
    void KilledUnit(Unit* /*victim*/) override
    {
        Talk(SAY_SLAY);
    }
```

**Branch(es):** 3.3.5
**TC rev. hash/commit:** 4ee6b67a84afb62b7c4b93e754b0dc5dd1ef8670

**Comment by jackpoz** [comment]
```c++
TalkKilledUnit(uint32 textid, bool checkCooldown = true, bool playerOnly = true);
```

something like this ?

**Comment by ccrs** [comment]
I'd go for a more modular solution, not just specific to KilledUnit

**Comment by offl** [comment]
>```c++
>TalkKilledUnit(uint32 textid, bool checkCooldown = true, bool playerOnly = true);
>```
>
>something like this ?

Yes but where we will define timers?

> I'd go for a more modular solution, not just specific to KilledUnit

In SAI we can add cooldown for nearly all event types(means to nearly all events from AI hooks like OnKill). But SAI is different
Original OnKill AI hook if I'm not wrong too has timers, I can't find right now the link with AI hooks and their params

But it sounds like a big rewrite

**Comment by ccrs** [comment]
SAI timers hare handled by the script system not the AI, so we cannot port anything there

**Comment by Ovahlord** [comment]
Just add a cooldown field to creature_text and and skip building chat messages when the id is on cooldown.

**Comment by HelloKitty** [comment]
Another possibility is making Talk events a first-class concept in the EventMap. Internally could be something like: `events.ScheduleTalk(SAY_ILLIDAN_KILL);`

These events could be serviced like EventMap::Update(uint32 time, Unit*) or maybe CreatureAI as parameter.

Internally this could call sCreatureTextMgr->SendChat or Talk. It's such a common event type that it really would be nice that EventMap provided handling for it rather than reimplementing the same logic in every ExecuteEvent. And this could apply to other simpler event types. EventMap could internally track time since last queued event, and not dispatch the event if it's too soon since the last talk.

Then the resulting code might look like:

``` 
void KilledUnit(Unit* victim) override
{
 if (victim->IsPlayer())
    events.ScheduleTalk(SAY_ILLIDAN_KILL);
}
```

I understand this is a radical proposal though hehe, and it's just an idea. Lots of ways to support the concept of Talk events and Cooldowns!

**Comment by offl** [comment]
> I understand this is a radical proposal though hehe, and it's just an idea.

To be honest that sounds better than involving creature_text.Duration because it's easier

**Comment by Nyr97** [comment]
But would this not mean that we would have to add as many events as talks we need to have on a script? Or it would call creature_text directly instead?

**Comment by ccrs** [comment]
maybe its overkill, but what about having a separate scheduler in scriptedcreature? maybe add a helper method too to ease the scheduling. this way we can have there everything that needs timed execution outside encounter script or whatever.

**Comment by Lordron** [comment]
it is possible that this cooldowns equals the duration of the sound tighted to the text. i saw this behaviour in 4 mop scenarios. could be extracted from sound files and stored somewhere in the db. you can download sound from wowhead and check (or see in wowhead audio player). For me it was a boss who say one text when he is on 1 hp, and then 3.3 seconds later another text right after first is complete PLUS instakill spell and other visuals

**Comment by Lordron** [comment]
ah sorry this is another issue

**Comment by offl** [comment]
>it is possible that this cooldowns equals the duration of the sound

No

**Comment by offl** [comment]
The problem is we script everything in C++ or everything in DB. Blizz scripts simple stuff in DB and hard stuff in Lua (their developer said that they used Lua for Onyxia's air phase). I think all texts are scripted in AIGroup (OnKill, OnDeath, OnAggro), DB scripting provides cooldown for events
The problem is we can't (?) use both systems at once and SAI is not good enough to start migrating all those texts to SAI

**Comment by Nyr97** [comment]
maybe a new vector storage for Creature class where we can store by pairs: talkId, current cooldown; iterate through it on UpdateAI to reduce current cooldown of existing pairs in the vector and prevent Talk from happening if the talkId's associated cooldown is higher than 0, maybe adding a new Optional member for Talk to set a given cooldown if set (this would push it to our new storage), then we can implement it on SAI under a new parameter

talk cooldowns are not exclusive to kill only, there are several mechanics which happen to have a cooldown for it to be pushed as well
